

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>General Usage &mdash; Metaboverse 0.0.0-alpha documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/metaboverse_logo.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="FAQs" href="faqs.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Metaboverse
          

          
            
            <img src="../_static/metaboverse_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Metaboverse</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">General Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#getting-started">Getting Started</a></li>
<li class="toctree-l2"><a class="reference internal" href="#regulatory-hotspot-identification-motif-analysis">Regulatory Hotspot Identification (Motif Analysis)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#general-visualization">General Visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#perturbation-network-modeling">Perturbation Network Modeling</a></li>
<li class="toctree-l2"><a class="reference internal" href="#nearest-neighborhood-searches">Nearest Neighborhood Searches</a></li>
<li class="toctree-l2"><a class="reference internal" href="#analyzing-previous-models">Analyzing Previous Models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#publishing-data-analyzed-with-metaboverse">Publishing Data Analyzed with Metaboverse</a></li>
<li class="toctree-l2"><a class="reference internal" href="#analysis-video-walkthrough">Analysis Video Walkthrough</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Metaboverse</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>General Usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/general-usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="general-usage">
<span id="general-link"></span><h1>General Usage<a class="headerlink" href="#general-usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="getting-started">
<h2>Getting Started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">1. If this is your first time analyzing a given dataset, you should press <code class="xref py py-data docutils literal notranslate"><span class="pre">Skip</span></code> on the opening page.</div>
</div>
<a class="reference internal image-reference" href="../_images/page_one.gif"><img alt="../_images/page_one.gif" class="align-center" src="../_images/page_one.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">2. Select the organism of interest you would like to model your data.</div>
<div class="line">3. Select an output location for files generated by Metaboverse.</div>
</div>
<a class="reference internal image-reference" href="../_images/page_two.gif"><img alt="../_images/page_two.gif" class="align-center" src="../_images/page_two.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">4. Provide the following variables for your model.</div>
<div class="line">a) Provide the appropriate data files for your -omics data. Please see the <a class="reference internal" href="overview.html#data-inputs"><span class="std std-ref">Data Inputs</span></a> sub-heading for more information.</div>
<div class="line">b) Provide a name for your experiment/model.</div>
<div class="line">c) Provide the appropriate experimental type. In this case, I am providing timecourse metabolomics data, so I will select that option.</div>
<div class="line">d) When processing a timecourse or multi-condition dataset, you will then be asked to provide the appropriate sample labels. These should be separated by commas, and should appear in the same order as listed in the dataframe.</div>
<div class="line">e) Select these modification options if desired.</div>
<div class="line-block">
<div class="line">- <em>Broadcast gene expression to proteins</em>: Check to broadcast gene expression values to proteins when protein values are not available (checked by default).</div>
<div class="line">- <em>Consider modifiers in reaction collapsing</em>: Check to include modifiers in reaction collapsing. Catalysts are included as outputs, inhibitors are included as inputs. Please refer to documentation for more information.</div>
</div>
<div class="line">f) Add or remove any items you wish to not be displayed in the network visualizations.</div>
</div>
<a class="reference internal image-reference" href="../_images/step_four.png"><img alt="../_images/step_four.png" class="align-center" src="../_images/step_four.png" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<a class="reference internal image-reference" href="../_images/page_three.gif"><img alt="../_images/page_three.gif" class="align-center" src="../_images/page_three.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">5. Build your network model. After your network model has been built, you will have the option to perform regulatory hotspot identification, model perturbation networks, or perform general visualization of pathways and components by clicking on the appropriate buttons at the bottom of the screen that will appear once Metaboverse has finished processing your data.</div>
</div>
<a class="reference internal image-reference" href="../_images/step_five.png"><img alt="../_images/step_five.png" class="align-center" src="../_images/step_five.png" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="regulatory-hotspot-identification-motif-analysis">
<h2>Regulatory Hotspot Identification (Motif Analysis)<a class="headerlink" href="#regulatory-hotspot-identification-motif-analysis" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Regulatory patterns can be searched for in the global reaction network using this page. Users can select different motif types by clicking on the appropriate buttons and adjusting their thresholds as necessary.</div>
<div class="line"><br /></div>
<div class="line">Currently, these motif types are available:</div>
<div class="line">1. <strong>Average</strong>: This operates by averaging the measured values on each side of a reaction (inputs vs outputs) and calculating the absolute difference. If that difference is greater than the threshold, that reaction will be registered as a motif.</div>
<div class="line">2. <strong>MaxMax</strong>: Compare the absolute maximum values between reactants and products that pass the provided threshold.</div>
<div class="line">3. <strong>MaxMin</strong>: Compare the absolute maximum and minimum values between reactants and products, or products and reactants, respectively, that pass the provided threshold.</div>
<div class="line">4. <strong>Sustained</strong>: Find instances of sustained perturbation along a reaction where there are both an input and an output that are not the same molecule with changes above the selected threshold.</div>
<div class="line">5. <strong>ModReg</strong>: Find instances with one regulated modifier and one core component in the reaction.</div>
<div class="line">6. <strong>TransReg</strong>: Find instances where a component is the same for input and output, is regulated, along with a modifier being regulated.</div>
<div class="line">7. <strong>PathMax</strong>: Compare the absolute maximum and minimum values across a pathway.</div>
<div class="line">8. <strong>PathCov</strong>: Compare total coverage of reactions with at least one measured component in pathways.</div>
<div class="line"><br /></div>
<div class="line">Users can then select a given motif, view the pathways that reaction is present in across the global reaction network, select a pathway to view, and see all other patterns of that type available in that pathway.</div>
</div>
<a class="reference internal image-reference" href="../_images/motif_search.gif"><img alt="../_images/motif_search.gif" class="align-center" src="../_images/motif_search.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">If the model includes timecourse or multi-condition data, the patterns present at each timepoint or condition can be viewed.</div>
</div>
<a class="reference internal image-reference" href="../_images/motif_time.png"><img alt="../_images/motif_time.png" class="align-center" src="../_images/motif_time.png" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Users can also choose to include the following modifications in the motif search:</div>
<div class="line">a) <strong>Use Modifiers</strong>: Check to include modifiers in motif analysis. Catalysts are included as outputs, inhibitors are included as inputs. You will need to re-run the motif to include modifiers.</div>
<div class="line">b) <strong>Exclude Hubs</strong>: Exclude high-hub components from consideration in the motif search. This will remove hubs with more than 100 connections. You will need to re-run the motif to include modifiers.</div>
</div>
<a class="reference internal image-reference" href="../_images/motif_options.png"><img alt="../_images/motif_options.png" class="align-center" src="../_images/motif_options.png" style="width: 250px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="general-visualization">
<h2>General Visualization<a class="headerlink" href="#general-visualization" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Users can explore classical pathways and all components of the network interactively by navigating to the <code class="xref py py-data docutils literal notranslate"><span class="pre">Visualization</span></code> page.</div>
<div class="line"><br /></div>
<div class="line">a) At the top of the legend, reference to the node types are displayed. Grey nodes are reactions, nodes with solid outlines are measured, and nodes with dashed outlines are inferred using the gene broadcasting feature. Components with significant statistical values based on the threshold (defined in section d) are bolded. Users can hover across the color scale to see what color corresponds to what value.</div>
<div class="line">- <strong>Motif reaction nodes will be enlarged and outlined in bold purple.</strong></div>
<div class="line">b) Relationship types between nodes are explained in this section of the legend.</div>
<div class="line">c) Users can toggle component, reaction, and expression labels on and off using the appropriate buttons. Users can also toggle gene nodes and compartment shading on and off with the appropriate buttons. Users can also press the <code class="xref py py-data docutils literal notranslate"><span class="pre">Save</span> <span class="pre">Graph</span></code> to select the viewed network to a <code class="xref py py-data docutils literal notranslate"><span class="pre">png</span></code> file. Users can decide whether to show collapsed reactions, or to show the full, non-collapsed pathway of reactions that would have been collapsed. Collapsed reactions are displayed by default. Users can also open the selected pathway in Reactome to view a more classical representation of the pathway.</div>
<div class="line">d) Users can modify the number of neighbors to plot when double-clicking on a node to expand its nearest reaction neighbors. Users can also select a threshold to not display any node with more than the defined number of connections. By modifying the statistic threshold, users can change the minimum statistical value needed to bold the component node for easy discrimination of significant measurements in the network.</div>
<div class="line">e) Metadata about pathways, reactions, and components, along with warnings or errors, will be displayed in this section of the legend.</div>
<div class="line">f) Users can select from the following super-menus:</div>
<div class="line">- <em>All entities</em>: A list of all components (metabolites, proteins, etc.)</div>
<div class="line">- <em>All pathways</em>: A list of all pathways</div>
<div class="line">- <em>Other</em>: A list of selected super-pathways that will group pathways in the next drop-down menu. For example, if selecting <code class="xref py py-data docutils literal notranslate"><span class="pre">Metabolism</span></code>, only pathways related to Metabolism will be shown in the next menu.</div>
<div class="line"><br /></div>
</div>
<a class="reference internal image-reference" href="../_images/vis_overview.png"><img alt="../_images/vis_overview.png" class="align-center" src="../_images/vis_overview.png" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Users can easily toggle between timepoints or conditions by moving the slider.</div>
</div>
<a class="reference internal image-reference" href="../_images/pathway_time.gif"><img alt="../_images/pathway_time.gif" class="align-center" src="../_images/pathway_time.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="perturbation-network-modeling">
<h2>Perturbation Network Modeling<a class="headerlink" href="#perturbation-network-modeling" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Users can explore the role and extent of perturbation within the reaction network in their model by navigating to the appropriate page. Users can modify the perturbation threshold for the fold change or statistical value.</div>
</div>
<a class="reference internal image-reference" href="../_images/perturbation_connectivity.gif"><img alt="../_images/perturbation_connectivity.gif" class="align-center" src="../_images/perturbation_connectivity.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="nearest-neighborhood-searches">
<h2>Nearest Neighborhood Searches<a class="headerlink" href="#nearest-neighborhood-searches" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Users interested in exploring a components reaction neighborhood can double-click the component to expand this view. The number of reaction neighbors can be selected, and limits on the number of connections a connecting component can have to be displayed can be modified. This hub selection is particularly helpful for connected components that are involved in several reactions.</div>
</div>
<a class="reference internal image-reference" href="../_images/neighbor_search.gif"><img alt="../_images/neighbor_search.gif" class="align-center" src="../_images/neighbor_search.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="analyzing-previous-models">
<h2>Analyzing Previous Models<a class="headerlink" href="#analyzing-previous-models" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Users can load a previously modeled network with their data on it by loading the <code class="xref py py-data docutils literal notranslate"><span class="pre">json</span></code> file output by Metaboverse when originally generated. This is done on the Home page.</div>
</div>
<a class="reference internal image-reference" href="../_images/load_json.gif"><img alt="../_images/load_json.gif" class="align-center" src="../_images/load_json.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Users can load a previously modeled organism network by loading the <code class="xref py py-data docutils literal notranslate"><span class="pre">pickle</span></code> file output by Metaboverse when originally generated. This is done on the Curation page.</div>
</div>
<a class="reference internal image-reference" href="../_images/load_pickle.gif"><img alt="../_images/load_pickle.gif" class="align-center" src="../_images/load_pickle.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="publishing-data-analyzed-with-metaboverse">
<h2>Publishing Data Analyzed with Metaboverse<a class="headerlink" href="#publishing-data-analyzed-with-metaboverse" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">When publishing analyses that used Metaboverse, we recommend attaching the appropriate <code class="xref py py-data docutils literal notranslate"><span class="pre">json</span></code> file that contains the network with your data overlaid as a supplementary file. This will allow for others to easily reproduce and explore your data.</div>
<div class="line"><br /></div>
<div class="line">Users can explore the metadata related to their Metaboverse model by clicking on the <code class="xref py py-data docutils literal notranslate"><span class="pre">Session</span> <span class="pre">Data</span></code> tab in the menu once the <code class="xref py py-data docutils literal notranslate"><span class="pre">json</span></code> file is loaded.</div>
</div>
<a class="reference internal image-reference" href="../_images/show_metadata.gif"><img alt="../_images/show_metadata.gif" class="align-center" src="../_images/show_metadata.gif" style="width: 700px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="analysis-video-walkthrough">
<h2>Analysis Video Walkthrough<a class="headerlink" href="#analysis-video-walkthrough" title="Permalink to this headline">¶</a></h2>
<iframe width="700" height="450" src="https://www.youtube.com/embed/ytTIlBKzq-c" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe><div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="faqs.html" class="btn btn-neutral float-right" title="FAQs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019-2020, Jordan A. Berg, et. al.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>